### PR TITLE
27-add damsel object

### PR DESCRIPTION
**Pull request by Sean-Nishi**
Added damsel class. Movement is based on both player and enemy1.

Had to edit the test world_map in level1.py to be able to display/test one damsel.

### CODE REVIEW

File: game/damsel.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+class Damsel(Entity):
+    """Damsel in distress that the player saves"""
+
+    def __init__(self, pos, groups, obstacle_sprites):
+        """Initialize a Damsel with level info
+
+        Each damsel is initialized with their starting position,
+        which sprite groups it is part of, boundary sprites,
+        and enemy/interactable sprites.
+        """
```
Comment (owalk) [inline game/damsel.py:21]: Missing params in type docs. See this example: [Example](https://github.com/Sean-Nishi/Lunk-Game/blob/16d8121e72f2ee0d36d74ed78954712a87d54bde/game/player.py#L204)

File: game/damsel.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        self.timer = 100
+
+        self.obstacle_sprites = obstacle_sprites
+
+        # starting position is facing down
+        self.direction.y = 1
+        self.status = "down"
+        self.import_damsel_assets()
+
+    def import_damsel_assets(self):
```
Comment (owalk) [inline game/damsel.py:48]: missing doc string

File: game/damsel.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            else:
+                # walk up
+                self.direction.y = -1
+        """
+        # admire player logic
+        """
+        if player.hitbox.centerx < 5*TILESIZE from self.hitbox.centerx:
+            self.speed = 0
+
+        """
```
Comment (owalk) [inline game/damsel.py:127]: These comments look messy and shouldn't get pushed to the main branch.
Can you stub the function with cleaner comments and add a TODO comment describing what needs to be done?
lines 84-127 should be cleaned up.
Can you add an issue to the backlog to finish this function later?

File: game/damsel.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        """Initialize a Damsel with level info
+
+        Each damsel is initialized with their starting position,
+        which sprite groups it is part of, boundary sprites,
+        and enemy/interactable sprites.
+        """
+
+        super().__init__(groups)
+
+        damselMovementsPath = "graphics/damsel/damselWalking.png"
```
Comment (owalk) [inline game/damsel.py:25]: unclear variable name. damselMovementImagePath is clearer

```Suggestion
        damselMovementImagePath = "graphics/damsel/damselWalking.png"
```

File: game/damsel.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        self.damselAnimations = SpriteSheet(damselMovementsPath)
```
Comment (owalk) [inline game/damsel.py:26]:
```Suggestion
        self.damselAnimations = SpriteSheet(damselMovementImagePath)
```

File: game/damsel.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        damselSelfImageRect = pygame.Rect(0, 0, SPRITE_WIDTH, SPRITE_HEIGHT)
+        self.colorKeyBlack = (0, 0, 0)
```
Comment (owalk) [inline game/damsel.py:28]: Magic numbers, the 0's can be constants with descriptive names. 
it's setting the RGB color key for [this](https://www.pygame.org/docs/ref/surface.html#pygame.Surface.set_colorkey) so have three constants, red, green, blue, and set them all to 0. 
```Suggestion
degreeOfRed = 0
degreeOfGreen = 0
degreeOfBlue = 0
self.colorKeyBlack = (degreeOfRed, degreeOfGreen, degreeOfBlue)
```

File: game/damsel.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        self.image = self.damselAnimations.image_at(
+            damselSelfImageRect, self.colorKeyBlack
+        )
+        self.rect = self.image.get_rect(topleft=pos)
+        # modify model rect to be a slightly less tall hitbox.
+        self.hitbox = self.rect.inflate(0, -10)
+        # self.mapSize = pygame.math.Vector2(map_size)
```
Comment (owalk) [inline game/damsel.py:35]: the commented code here is unclear why it is commented out and still left in to push to main.
change to a descriptive TODO or remove the comment

File: game/damsel.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            "down": self.damselAnimations.load_strip(
+                walkingDownRect, 3, self.colorKeyBlack
+            ),
+            "left": self.damselAnimations.load_strip(
+                walkingLeftRect, 3, self.colorKeyBlack
+            ),
+            "right": self.damselAnimations.load_strip(
+                walkingRightRect, 3, self.colorKeyBlack
+            ),
+            "up_idle": [],
```
Comment (owalk) [inline game/damsel.py:68]: we should wait on adding additional animations until they will be actively used. lines 68-75 are not needed.

File: game/damsel.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            "down_attack": [],
+            "left_attack": [],
+            "right_attack": [],
+        }
+
+    def move(self, speed):
+        """Method to handle movement
+
+        Movement logic is as described in documentation.
+        """
```
Comment (owalk) [inline game/damsel.py:82]: Missing speed param in doc strings. See this example: [Example](https://github.com/Sean-Nishi/Lunk-Game/blob/16d8121e72f2ee0d36d74ed78954712a87d54bde/game/player.py#L204)

File: game/damsel.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            if seed % 3:
+                self.direction.y = -1
+            else:
+                self.direction.y = 1
+            self.timer = 0
+
+        # prevent diagonal moving from increasing speed
+        if self.direction.magnitude() != 0:
+            self.direction = self.direction.normalize()
+        # update
```
Comment (owalk) [inline game/damsel.py:150]: unclear comment. update what?

File: game/damsel.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        if direction == "y":
+            return pygame.transform.rotate(image, self.direction.y)
+        if direction == "x":
+            return pygame.transform.rotate(image, self.direction.x)
+
+    def animate(self):
+        """Method to loop through damsel animations"""
+        animation = self.animations[self.status]
+
+        # loop over fram index
```
Comment (owalk) [inline game/damsel.py:202]: frame has a typo
```Suggestion
        # loop over frame index
```

File: game/damsel.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        if self.direction.x > 0 and self.status != "right":
+            self.status = "right"
+        elif self.direction.x < 0 and self.status != "left":
+            self.status = "left"
+        if self.direction.y > 0 and self.status != "down":
+            self.status = "down"
+        elif self.direction.y < 0 and self.status != "up":
+            self.status = "up"
+
+    def set_image_direction(self, image):
```
Comment (owalk) [inline game/damsel.py:174]: Missing image param in doc strings. See this example: [Example](https://github.com/Sean-Nishi/Lunk-Game/blob/16d8121e72f2ee0d36d74ed78954712a87d54bde/game/player.py#L204)

**Review by owalk — CHANGES_REQUESTED** [review]

**Comment by owalk** [comment]
@Sean-Nishi just a heads up, can you wait to resolve the comments until I review everything again so it is easier to check that everything gets addressed

**Comment by owalk** [comment]
Can you 'fixup' the new commit so that changes are not being done to the code you introduced in the same PR?
git rebase, and then select fixup to the second commit to merge it with the first

File: game/damsel.py (hunk 6 above)
Comment (Sean-Nishi) [reply game/damsel.py:55]: If you do not want magic numbers, I suggest you change your code on line 27 in player.py to reflect this standard. I was following what you were doing/what was approved already.

File: game/damsel.py (hunk 8 above)
Comment (Sean-Nishi) [reply game/damsel.py]: Again, I was using what you added to player.py, lines 70-77, as the standard for the animations. Let me know if I can keep it as is or if you will change your code to match.

Comment (oliver-walker-hms) [reply game/damsel.py]: The rest of the code should match this as well, we can add it to the backlog as a small task to do later

File: game/damsel.py (hunk 6 above)
Comment (owalk) [reply game/damsel.py:55]: I added a new task [here](https://github.com/Sean-Nishi/Lunk-Game/issues/94) for this

File: game/damsel.py (hunk 8 above)
Comment (owalk) [reply game/damsel.py]: I added a new task for this [Here](https://github.com/Sean-Nishi/Lunk-Game/issues/95)

File: game/damsel.py (hunk 1 above)
Comment (owalk) [reply game/damsel.py:21]: This was missed

File: game/damsel.py (hunk 6 above)
Comment (owalk) [reply game/damsel.py:59]: This was missed. If we want to go back to it later can you update the backlog task to reference this file and the appropriate line number?

File: game/player.py [hunk 13]
```diff
@@ -215,7 +215,7 @@ def collision_check(self, direction):
                 if sprite.hitbox.colliderect(self.hitbox):
                     lastDirectionOfCollision = [self.direction.x, self.direction.y]
                     # reverse direction
-                    self.direction.x = -lastDirectionOfCollision.x
+                    self.direction.x = -lastDirectionOfCollision[0]
```
Comment (owalk) [inline game/player.py:218]: changes to player are out of scope for this commit, can you separate these changes into a new commit?
It looks like the branch is not up to date with main, this line on main looks like the following:
```
self.direction.x *= -1
 ```
You should probably rebase to the previous commit and do a git fetch, followed by a git reset --hard origin/main

File: game/player.py [hunk 14]
```diff
@@ -224,7 +224,7 @@ def collision_check(self, direction):
                 if sprite.hitbox.colliderect(self.hitbox):
                     lastDirectionOfCollision = [self.direction.x, self.direction.y]
                     # reverse direction
-                    self.direction.y = -lastDirectionOfCollision.y
+                    self.direction.y = -lastDirectionOfCollision[1]
```
Comment (owalk) [inline game/player.py:227]: same comment as above, out of scope for the commit and getting the branch inline with the main branch will overwrite these changes

**Review by owalk — CHANGES_REQUESTED** [review]

**Review by owalk — APPROVED** [review]

**Review by owalk — APPROVED** [review]

**Pull request merged**